gapis/api/vulkan/framegraph.go

```diff
@@ -16,14 +16,153 @@ package vulkan
 
 import (
 	"context"
+	"fmt"
 
-	"github.com/google/gapid/core/log"
 	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/sync"
+	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/resolve/dependencygraph2"
 	"github.com/google/gapid/gapis/service/path"
 )
 
+// renderpassInfo stores a renderpass' info relevant for the framegraph.
+type renderpassInfo struct {
+	id       uint64
+	beginIdx api.SubCmdIdx
+	endIdx   api.SubCmdIdx
+	nodes    []dependencygraph2.NodeID
+	deps     map[uint64]struct{} // set of renderpasses this renderpass depends on
+}
+
 // GetFramegraph creates the framegraph of the given capture.
 func (API) GetFramegraph(ctx context.Context, p *path.Capture) (*api.Framegraph, error) {
-	log.W(ctx, "TODO: implement framegraph creation")
-	return nil, nil
+	config := dependencygraph2.DependencyGraphConfig{
+		SaveNodeAccesses:    true,
+		ReverseDependencies: true,
+	}
+	dependencyGraph, err := dependencygraph2.GetDependencyGraph(ctx, p, config)
+	if err != nil {
+		return nil, err
+	}
+
+	rpInfos := []*renderpassInfo{}
+	var rpInfo *renderpassInfo
+	currRpId := uint64(0)
+
+	// postSubCmdCb effectively processes each subcommand to extract renderpass info.
+	postSubCmdCb := func(state *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd, i interface{}) {
+		vkState := GetState(state)
+		cmdRef := i.(CommandReferenceʳ)
```

Comment:
Do we know that i has a type of `CommandReferenceʳ` ? If yes, we should replace it with `interface{}`? if no, then we should check if type conversion is valid. Otherwise it's a direct crash.

Reply:
Added a type check.